File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    let complements = {
        ")":"(",
        "}":"{",
        "]":"["
    }
    
    for(let c of s){
        if(c in complements){
            if(stack.length && stack[stack.length - 1] === complements[c]){
                stack.pop();
            } else {
                return false;
            }
        } else {
            stack.push(c);
        }
    }
    
    if(!stack.length){
        return true;
    } else {
        return false;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        complements = {
            ")" : "(",
            "}" : "{",
            "]" : "["
        }
        
        stack = []
        
        for c in s:
            if c in complements:
                if stack and stack[-1] == complements[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        
        if stack:
            return False
        else:
            return True
        